modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/HeapLockManager.java

```diff
@@ -328,10 +357,331 @@ private LockState adjustLockState(LockState state, LockState v) {
+
+                            return completedFuture(lock);
+                        }
+                    } finally {
+                        stripedLock.readLock(idx).unlock();
+                    }
+
+                default:
+                    assert false : "Unsupported coarse lock mode: " + lockMode;
+            }
+
+            return null; // Should not be here.
+        }
+
+        /**
+         * Releases the lock. Should be called from {@link #releaseAll(UUID)}.
+         *
+         * @param lock The lock.
+         */
+        public void release(@Nullable Lock lock) {
+            if (lock == null) {
+                return;
+            }
+
+            switch (lock.lockMode()) {
+                case S:
+                    stripedLock.writeLock().lock();
+
+                    try {
+                        Lock removed = slockOwners.remove(lock.txId());
+
+                        assert removed != null : "Attempt to release not acquired lock: " + lock.txId();
+                    } finally {
+                        stripedLock.writeLock().unlock();
+                    }
+
+                    break;
+                case IX:
+                    int idx = Math.floorMod(spread(lock.txId().hashCode()), CONCURRENCY);
+
+                    Map<UUID, IgniteBiTuple<Lock, CompletableFuture<Lock>>> wakeups;
+
+                    stripedLock.readLock(idx).lock();
+
+                    try {
+                        var removed = ixlockOwners.remove(lock.txId());
+
+                        assert removed != null : "Attempt to release not acquired lock: " + lock.txId();
+
+                        if (slockWaiters.isEmpty()) {
+                            return; // Nothing to do.
+                        }
+
+                        if (!ixlockOwners.isEmpty()) {
+                            assert slockOwners.isEmpty() || slockOwners.containsKey(lock.txId());
+
+                            return; // Nothing to do.
+                        }
+
+                        // No race here because no new locks can be acquired after releaseAll due to 2-phase locking protocol.
+
+                        // Promote waiters to owners.
+                        wakeups = new HashMap<>(slockWaiters);
+
+                        slockWaiters.clear();
+
+                        for (IgniteBiTuple<Lock, CompletableFuture<Lock>> value : wakeups.values()) {
+                            slockOwners.put(value.getKey().txId(), value.getKey());
+                        }
+                    } finally {
+                        stripedLock.readLock(idx).unlock();
+                    }
+
+                    for (Entry<UUID, IgniteBiTuple<Lock, CompletableFuture<Lock>>> entry : wakeups.entrySet()) {
+                        entry.getValue().get2().complete(entry.getValue().get1());
+                    }
+
+                    break;
+                default:
+                    assert false : "Unsupported coarse unlock mode: " + lock.lockMode();
```

Comment:
seems SIX lock mode is also possible here

Reply:
No, in the current implementation SIX lock is acquired by holding two locks: IX + S.

Reply:
Agree, I should have checked more carefully.

Reply:
Agree, I should have checked more carefully.